### PR TITLE
Fixes #1983: Created textSearch actions, actionTypes, reducers

### LINKED ISSUE #1983: Create action, reducer and actionType for message search ( MessageSection )
**Actual Behaviour**

The search state is being stored in the MessageSection itself which is making it hard for the TopBar to be reused

**Expected Behaviour**

Shift this state and the functions used to manipulate it to the redux store

**Would you like to work on the issue?**

yes

### PR DESCRIPTION
**Pull request by AakashMallik**
Fixes #1983 

Changes: The change I have made are:
- [x] Create actionTypes for text search
- [x] Create actions for text Search
- [x] Completly implement the reducers for text search.

Demo Link: https://pr-1994-fossasia-susi-web-chat.surge.sh

Note: I have left a few comments in the reducer intentionally as that will help me while I am working on the integration of actions in the component. I'll remove them in the upcomming PR.

### CODE REVIEW

**Comment by AakashMallik** [comment]
@akshatnitd @fragm3 A review would be nice.

File: src/redux/reducers/searchText.js [hunk 1]
```diff
@@ -0,0 +1,198 @@
+import { handleActions } from 'redux-actions';
+import actionTypes from '../actionTypes';
+import { searchMessages } from '../../utils/searchMessages';
+
+const defaultState = {
+  search: false,
```
Comment (akshatnitd) [inline src/redux/reducers/searchText.js]: What does `search` signify? Can you give it a more apt name like `isSearchBarActive` depending on whatever it means.

File: src/redux/reducers/searchText.js [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+  searchState: {
+    markedMessagesByID: {},
```
Comment (akshatnitd) [inline src/redux/reducers/searchText.js]: This state would not be required now. We will already have message by ids object in our store

File: src/redux/reducers/searchText.js [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+};
+
+export default handleActions(
+  {
+    [actionTypes.SEARCH_TEXT_CHANGED](state, { payload }) {
+      // TODO: Call close snackbar from Top Bar
+      const { messages, messagesByID, searchText } = payload;
+      let { searchState } = state;
+      // const searchText = event.target.value;
+      if (searchText.trim()) {
```
Comment (akshatnitd) [inline src/redux/reducers/searchText.js]: We should consider that the searchText received in the reducer is already trimmed

File: src/redux/reducers/searchText.js [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+          markedIndices: markedIndices,
+          scrollLimit: markedIDs.length,
+          scrollIndex: 0,
+          scrollID: markedIDs[0],
+          caseSensitive: searchState.caseSensitive,
+          searchIndex: 1,
+          searchText: searchText,
+        };
+        if (markedIDs.length === 0 && searchText.trim()) {
+          // if no Messages are marked(i.e no result) and the search query is not empty
```
Comment (akshatnitd) [inline src/redux/reducers/searchText.js]: This functionality can be achieved by adding a new state to searchState

File: src/redux/reducers/searchText.js [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    },
+    [actionTypes.SEARCH_PREVIOUS_ITEM](state, { payload }) {
+      // TODO: Find a way to pass ref for ul
+      let { ul } = payload;
+      const { searchState } = state;
+      let newIndex = searchState.scrollIndex + 1;
+      let newSearchCount = searchState.searchIndex + 1;
+      let indexLimit = searchState.scrollLimit;
+      let markedIDs = searchState.markedIDs;
+      if (newSearchCount <= 0) {
```
Comment (akshatnitd) [inline src/redux/reducers/searchText.js]: Is this condition required

**Comment by AakashMallik** [comment]
@akshatnitd I'll make the changes as soon as I am done with my exams. Thanks for reviewing my work. :smile:

File: src/redux/reducers/searchText.js (hunk 2 above)
Comment (AakashMallik) [reply src/redux/reducers/searchText.js]: @akshatnitd True, but sir this is being used to maintain a copy of the message object that is being modified by the **SearchMessage** util

```
const orignalMessageText = messagesByID[id].text;
    let messageCopy = { ...messagesByID[id] };
    let messageText = orignalMessageText;
    if (orignalMessageText) {
      if (!isCaseSensitive) {
        searchText = searchText.toLowerCase();
        messageText = messageText.toLowerCase();
      }
      const match = messageText.indexOf(searchText);
      if (match !== -1) {
        messageCopy.mark = {
          searchText,
          isCaseSensitive,
        };
        markingData.markedIDs = [messageCopy.id, ...markingData.markedIDs];
        markingData.markedIndices = [index, ...markingData.markedIndices];
      }
      markingData.markedMessagesByID[id] = messageCopy;
```

Comment (AakashMallik) [reply src/redux/reducers/searchText.js]: Any suggestions, sir? @akshatnitd

File: src/redux/reducers/searchText.js (hunk 5 above)
Comment (AakashMallik) [reply src/redux/reducers/searchText.js]: @akshatnitd I'll work on optimizing prev and next search in this PR, I'll try to get rid of redundancies.

File: src/redux/reducers/searchText.js (hunk 4 above)
Comment (AakashMallik) [reply src/redux/reducers/searchText.js]: @akshatnitd I'll have to think of some creative way to optimize searchMessage Util, maybe then I can get rid of redundant variables in this reducer. It is currently dependent on what the util is returning. So, I'll rater open a new issue for searchMessage util optimization.

**Pull request closed without merging**